File: AlgoExpert/Easy/Branch_Sums/CPP/Branch_Sums.cpp
/*
Problem Name: Branch Sums
Difficulty: Easy
Platform: AlgoExpert

Problem Description:
Write a function that takes in a Binary Tree and returns a list of its branch sums ordered
from the leftmost branch sum to the rightmost.

Time Complexity: O(n)
Space Complexity: O(n)
*/

#include <iostream>
#include <vector>
using namespace std;

class BinaryTree {
public:
    int value;
    BinaryTree* left;
    BinaryTree* right;

    BinaryTree(int value) {
        this->value = value;
        left = nullptr;
        right = nullptr;
    }
};

void calculateBranchSums(BinaryTree* node, int runningSum, vector<int>& sums) {
    if (node == nullptr) return;

    int newRunningSum = runningSum + node->value;

    if (node->left == nullptr && node->right == nullptr) {
        sums.push_back(newRunningSum);
        return;
    }

    calculateBranchSums(node->left, newRunningSum, sums);
    calculateBranchSums(node->right, newRunningSum, sums);
}

vector<int> branchSums(BinaryTree* root) {
    vector<int> sums;
    calculateBranchSums(root, 0, sums);
    return sums;
}

File: AlgoExpert/Easy/Branch_Sums/Python/Branch_Sums.py
"""
Problem Name: Branch Sums
Difficulty: Easy
Platform: AlgoExpert

Problem Description:
Return a list of sums of all values from root to each leaf node in a binary tree.

Time Complexity: O(n)
Space Complexity: O(n)
"""

class BinaryTree:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


def branchSums(root):
    sums = []
    calculateBranchSum(root, 0, sums)
    return sums

def calculateBranchSum(node, runningSum, sums):
    if node is None:
        return
    newRunningSum = node.value + runningSum
    if node.right is None and node.left is None:
        sums.append(newRunningSum)
        return
    calculateBranchSum(node.left, newRunningSum, sums)
    calculateBranchSum(node.right, newRunningSum, sums)

